src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/_validators.py

```diff
@@ -321,6 +321,99 @@ def validate_select(namespace):
+
+    if storage_desc.blob is not None:
+        raise ValueError('incorrect usage: --source should be either container URL or name')
+
+    if namespace.account_name is None:
+        namespace.account_name = storage_desc.account
+
+    namespace.source_container_name = storage_desc.container
+
+
+def process_upload_batch_parameters(namespace):
+    """Process the source and destination of storage blob upload command"""
+    from azure.cli.command_modules.storage.storage_url_helpers import parse_url
+
+    # 1. quick check
+    if namespace.destination is None:
+        raise ValueError('incorrect usage: --destination is missing.')
+
+    if namespace.source is None:
+        raise ValueError('incorrect usage: --source is missing.')
+
+    if not os.path.exists(namespace.source):
+        raise ValueError('incorrect usage: source {} does not exist'.format(namespace.source))
+
+    if not os.path.isdir(namespace.source):
+        raise ValueError('incorrect usage: source must be a directory')
+
+    # 2. try to extract account name and container name from destination string
+    storage_desc = parse_url(namespace.destination)
+
+    if storage_desc.blob is not None:
+        raise ValueError('incorrect usage: destination cannot be a blob url')
+
+    namespace.destination_container_name = storage_desc.container
+
+    if not namespace.account_name:
+        namespace.account_name = storage_desc.account
+
+    # 3. collect the files to be uploaded
+    namespace.source = os.path.realpath(namespace.source)
+
+    if namespace.pattern:
+        from fnmatch import fnmatch
+        pattern = os.path.join(namespace.source, namespace.pattern.lstrip('/'))
+
+        def _file_selector(file_path):
+            return fnmatch(file_path, pattern)
+    else:
+        def _file_selector(_):
+            return True
+
+    source_files = []
+
+    from os import walk
+    for root, _, files in walk(namespace.source):
+        #pylint: disable=bad-builtin
+        filtered = filter(_file_selector, (os.path.join(root, f) for f in files))
+        source_files += [(path, path[len(namespace.source) + 1:]) for path in filtered]
+
+    namespace.source_files = source_files
+
+    if namespace.blob_type is None:
+        vhd_files = [f for f in source_files if f[0].endswith('.vhd')]
+        if any(vhd_files) and len(vhd_files) == len(source_files):
+            # when all the listed files are vhd files use page
+            namespace.blob_type = 'page'
+        elif any(vhd_files):
+            # source files contain vhd files but not all of them
+            raise ValueError('''Fail to guess the required blob type. Type of the files to be
+            uploaded are not consistent. Default blob type for .vhd files is "page", while
+            others are "block". You can solve this problem by either explicitly set the blob
+            type or ensure the pattern matches a correct set of files.''')
+        else:
+            namespace.blob_type = 'block'
```

Comment:
Alternatively you could simply dynamically set blob_type for each copy operation based on the file type, or split them into VHDs and "all others" instead of erroring out. I'm fine with the error though.